tcmur_work.c

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2020 Red Hat, Inc.
+ *
+ * This file is licensed to you under your choice of the GNU Lesser
+ * General Public License, version 2.1 or any later version (LGPLv2.1 or
+ * later), or the Apache License 2.0.
+ */
+#define _GNU_SOURCE
+#include <stdlib.h>
+#include <stdint.h>
+#include <errno.h>
+#include <pthread.h>
+
+#include "libtcmu.h"
+#include "libtcmu_log.h"
+#include "tcmur_device.h"
+#include "tcmur_work.h"
+
+struct tcmur_work *tcmur_create_work(void)
+{
+	struct tcmur_work *work;
+
+	work = calloc(1, sizeof(*work));
+	if (!work)
+		return NULL;
+
+	if (pthread_mutex_init(&work->lock, NULL))
+		goto free_work;
+
+	if (pthread_cond_init(&work->cond, NULL))
+		goto destroy_mutex;
+
+	return work;
+
+destroy_mutex:
+	pthread_mutex_destroy(&work->lock);
+free_work:
+	free(work);
+	return NULL;
+}
+
+static void __tcmur_flush_work(struct tcmur_work *work)
+{
+	/*
+	 * The lock thread may need to do a handler reopen call and try to flush
+	 * itself. Just ignore.
+	 */
+	if (pthread_self() == work->thread)
+		return;
+
+	/*
+	 * Some handlers will crash if we do a cancel so we just wait.
+	 */
+	tcmu_dbg("waiting for %lu to complete\n", work->thread);
+	while (work->refcnt)
+		pthread_cond_wait(&work->cond, &work->lock);
+}
+
+static void *tcmur_work_fn(void *data)
+{
+	struct tcmur_work *work = data;
+
+	work->work_fn(work->data);
+
+	pthread_mutex_lock(&work->lock);
+	if (--work->refcnt == 0)
+		pthread_cond_signal(&work->cond);
+	pthread_mutex_unlock(&work->lock);
+	return NULL;
+}
+
+void tcmur_flush_work(struct tcmur_work *work)
+{
+	pthread_mutex_lock(&work->lock);
+	__tcmur_flush_work(work);
+	pthread_mutex_unlock(&work->lock);
+}
+
+int tcmur_run_work(struct tcmur_work *work, void *data, void (*work_fn)(void *))
+{
+	pthread_attr_t attr;
+	int ret;
+
+	pthread_attr_init(&attr);
+	pthread_attr_setdetachstate(&attr, PTHREAD_CREATE_DETACHED);
+
+	pthread_mutex_lock(&work->lock);
+	work->work_fn = work_fn;
+	work->data = data;
+	ret = pthread_create(&work->thread, &attr, tcmur_work_fn, work);
```

Comment:
This would stomp on the `work->thread` of a previously running thread, right?  That makes `__tcmur_flush_work` check of the thread suspicious.

Reply:
Yeah, right.

I have removed the thread in work and will use the pthread name to identify whether the work thread itself is trying to call the tcmur_flush_work(), which will cause stuck forever.

Reply:
Together with this versoin, I have remove the lock_work and since we can run multiple thread by using one event_work. And also remove the event type when in event reporting helper, since we are always reporting all the event every time.